bin/setupReplication.js

```diff
@@ -0,0 +1,431 @@
+            },
+        ],
+    };
+}
+
+function _setupS3Client(host, port, profile) {
+    const credentials = new SharedIniFileCredentials({ profile });
+    return new S3({
+        endpoint: `http://${host}:${port}`,
+        sslEnabled: false,
+        credentials,
+        s3ForcePathStyle: true,
+    });
+}
+
+
+function _setupIAMClient(host, port, profile) {
+    const credentials = new SharedIniFileCredentials({ profile });
+    return new IAM({
+        endpoint: `http://${host}:${port}`,
+        sslEnabled: false,
+        credentials,
+        maxRetries: 0,
+        region: 'us-east-1',
+        signatureCache: false,
+        s3ForcePathStyle: true,
+        httpOptions: {
+            timeout: 1000,
+        },
+    });
+}
+
+class _SetupReplication {
+    /**
+     * This class sets up two buckets for replication.
+     * @constructor
+     * @param {String} sourceBucket - Source Bucket Name
+     * @param {String} targetBucket - Destination Bucket Name
+     * @param {Object} log - Werelogs Request Logger object
+     * @param {Object} config - bucket configurations
+     */
+    constructor(sourceBucket, targetBucket, log, config) {
+        const { source, destination } = config.extensions.replication;
+        this._log = log;
+        this._sourceBucket = sourceBucket;
+        this._targetBucket = targetBucket;
+        this._s3Clients = {
+            source: _setupS3Client(source.s3.host, source.s3.port,
+                'backbeatsource'),
+            target: _setupS3Client(destination.s3.host, destination.s3.port,
+                'backbeattarget'),
+        };
+        this._iamClients = {
+            source: _setupIAMClient(source.auth.vault.host,
+                 source.auth.vault.iamPort,
+                'backbeatsource'),
+            target: _setupIAMClient(destination.auth.vault.host,
+                 destination.auth.vault.iamPort,
+                'backbeattarget'),
+        };
+    }
+
+    _checkSanity(cb) {
+        return async.waterfall([
+            next => this._isValidBucket('source', next),
+            next => this._isValidBucket('target', next),
+            next => this._isVersioningEnabled('source', next),
+            next => this._isVersioningEnabled('target', next),
+            next => this._isReplicationEnabled('source', next),
+            (srcArn, next) => this._isValidRole('source', srcArn, next),
+            (tgtArn, next) => this._isValidRole('target', tgtArn, next),
+        ], cb);
+    }
+
+    _isValidBucket(where, cb) {
+        // Does the bucket exist and is it reachable?
+        const bucket = where === 'source' ? this._sourceBucket :
+            this._targetBucket;
+        this._s3Clients[where].headBucket({ Bucket: bucket }, err => {
+            if (err) {
+                this._log.error('bucket sanity check error', {
+                    error: err,
+                    method: '_SetupReplication._isValidBucket',
+                    errorMessage: err.message,
```

Comment:
For error logging, I followed https://github.com/scality/S3/pull/845#discussion_r131225448
Not too sure which way to do this, going to wait on this one until we get a more definitive answer